server/src/fstar_connection.ts

```diff
@@ -0,0 +1,368 @@
+		else
+			console.warn("Could not retrieve the" + stream + "stream for this F* proccess");
+	}
+
+	// Custom events that will be emitted on top of 'data' events. These events
+	// can be listened for.
+	private addCustomEvents() {
+		const stdout_stream = this.fstar.proc.stdout;
+		if (!stdout_stream) {
+			console.warn("Could not retrieve the stdout stream to register custom event handlers");
+			return;
+		}
+		const handler = (msg: object) => this.emitPerMessage(stdout_stream, msg);
+		// The bufferHandler buffers up received input until it finds a valid
+		// message. The wrapped handler will then be called with the parsed
+		// valid message. This handles receiving fragmented messages or multiple
+		// messages over the stream from the F* process.
+		const bufferHandler = FStarConnection.bufferedMessageHandlerFactory(handler);
+		// Note: listeners stack rather than overwrite each other so other
+		// 'data' handlers won't overwrite our custom events handler.
+		stdout_stream.on('data', bufferHandler);
+	}
+
+	// Emits the appropriate events for a given message. This helper adds the
+	// following events:
+	// - 'message': emitted for each valid F* message, emits the message which
+	//   is a valid javascript object.
+	// - 'message:protocol-info': emitted for each protocol-info message, emits
+	//   the contents as a `ProtocolInfo` object.
+	// - 'message:ide-progress': emitted for each progress message, emits the
+	//   contents as an `IdeProgress` object.
+	// - 'message:ide-proof-state': emitted for each proof-state message, emits
+	//   the contents as an `IdeProofState` object.
+	// - 'message:ide-info': emitted for each info message, emits the contents
+	//   as-is.
+	// - 'message:ide-symbol': emitted for each symbol message, emits the
+	//   response as an `IdeSymbol` object.
+	// - 'message:ide-error': emitted for each error message, emits the response
+	//   as an `IdeError[]` object.
+	// - 'message:ide-auto-complete': emitted for each auto-complete message,
+	//   emits the response as an `IdeAutoCompleteResponses` object.
+	private emitPerMessage(stream: Readable, msg: object) {
+		stream.emit('message', msg);
+
+		// Events for specific message types
+		if (isProtocolInfo(msg)) {
+			stream.emit('message:protocol-info', msg as ProtocolInfo);
+		} else if (isIdeQueryResponse(msg)) {
+			const r = msg as IdeQueryResponse;
+			if (r.kind === 'message' && r.level === 'progress') {
+				stream.emit('message:ide-progress', r.contents as IdeProgress);
+			} else if (r.kind === 'message' && r.level === 'proof-state') {
+				stream.emit('message:ide-proof-state', r.contents as IdeProofState);
+			} else if (r.kind === 'message' && r.level === 'info') {
+				stream.emit('message:ide-info', r.contents);
+			} else if (r.kind === "response" && r.response) {
+				const responseType = this.decideIdeReponseType(r.response);
+				if (responseType === 'symbol') {
+					stream.emit('message:ide-symbol', r.response as IdeSymbol);
```

Comment:
I'm not a huge fan of untyped callbacks.  The untyped part for obvious reasons, but also the callback part because it leads to code where the request and the response handling are completely disconnected.

Ideally the `FStarConnection` level API would be an async function:
```
async lookupQuery(position: Position, word: string, range: FStarRange): Promise<IdeSymbol>;
```

(In an ideal world, this function could be built on top of a generic `async request(query: string, args: any): Promise<{kind: string, status: string, response: any}>` function.  But writing such a generic function is a bit complicated in F*, since some requests have no response message, some have exactly one, and some have more than one.)

Reply:
Filed as #27.